rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,326 @@
+    /// # Panics
+    /// Possible panic if, for some error, the fencing token (a.k.a. `version`) of the acquired lease is None.
+    pub async fn acquire(
+        &mut self,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+        renewal_period: Option<Duration>,
+    ) -> Result<HybridLogicalClock, Error> {
+        // Stop auto-renewal.
+        if !self.auto_renewal_cancellation_token.is_cancelled() {
+            self.auto_renewal_cancellation_token.cancel();
+        }
+
+        let acquire_result = self
+            .internal_acquire(lease_expiration, request_timeout)
+            .await;
+
+        if renewal_period.is_some() {
+            self.auto_renewal_cancellation_token = CancellationToken::new();
+            let mut self_clone = self.clone();
+
+            tokio::task::spawn({
+                async move {
+                    loop {
+                        select! {
+                            () = self_clone.auto_renewal_cancellation_token.cancelled() => {
+                                break; // Auto-renewal is cancelled.
+                            }
+                            () = tokio::time::sleep(renewal_period.unwrap()) => {}
+                        }
+
+                        if self_clone
+                            .internal_acquire(lease_expiration, request_timeout)
+                            .await
+                            .is_err()
+                        {
+                            // Acquire failed. Stopping Auto-renewal.
+                            break;
+                        }
+                    }
+                }
+            });
+        }
+
+        acquire_result
+    }
+
+    /// Releases a lease if and only if requested by the lease holder (same client id).
+    ///
+    /// Note: `request_timeout` is rounded up to the nearest second.
+    ///
+    /// Returns `Ok()` if lease is no longer held by this `lease_holder`, or `Error` otherwise.
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidArgument`](ErrorKind::InvalidArgument) if the `request_timeout` is zero or > `u32::max`
+    ///
+    /// [`struct@Error`] of kind [`ServiceError`](ErrorKind::ServiceError) if the State Store returns an Error response
+    ///
+    /// [`struct@Error`] of kind [`UnexpectedPayload`](ErrorKind::UnexpectedPayload) if the State Store returns a response that isn't valid for a `V Delete` request
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if there are any underlying errors from the command invoker
+    ///
+    /// # Panics
+    /// Possible panic if `std::Lock::lock()` fails.
+    pub async fn release(&mut self, request_timeout: Duration) -> Result<(), Error> {
+        // Stop auto-renewal.
+        if !self.auto_renewal_cancellation_token.is_cancelled() {
+            self.auto_renewal_cancellation_token.cancel();
+        }
+
+        *self
```

Comment:
should this only happen if the state store delete was successful?

Reply:
I think this makes sense - even if the operation wasn't successful, you still don't want to be holding the lease anymore. So you'd probably re-try this operation, but you wouldn't want any of your other code to try and use the fencing token. I think my concern was holding different state than the state store, but you're right that we don't have any super strong guarantee that the lease is still held in state store, so it's better to be conservative and assume we don't. Calling acquire when not technically needed won't be a harmful operation.
(keeping this item open just to track adding the doc, please resolve this comment once you've done that!)